nncf/experimental/torch/nas/bootstrapNAS/search/search.py

```diff
@@ -0,0 +1,386 @@
+                if dim == ElasticityDim.KERNEL:
+                    sample[dim] = [self._search._kernel_search_space[j - start_index][x[i][j]] for j in
+                                   range(start_index, start_index + len(self._search._kernel_search_space))]
+                    start_index += len(self._search._kernel_search_space)
+                elif dim == ElasticityDim.WIDTH:
+                    sample[dim] = {key - start_index: self._search._width_search_space[key - start_index][x[i][key]] for
+                                   key in range(start_index, start_index + len(self._search._width_search_space))}
+                    start_index += len(self._search._width_search_space)
+                elif dim == ElasticityDim.DEPTH:
+                    sample[dim] = self._search._valid_depth_configs[x[i][start_index]]
+                    start_index += 1
+
+            self._elasticity_handler.set_config(sample)
+
+            if sample != self._elasticity_handler.get_active_config():
+                nncf_logger.warning("Requested configuration was invalid")
+                nncf_logger.warning(f"Requested: {sample}")
+                nncf_logger.warning(f"Provided: {self._elasticity_handler.get_active_config()}")
+                self._search._bad_requests.append((sample, self._elasticity_handler.get_active_config()))
+
+            result = []
+            for key in sample.keys():
+                result.append(key.value)
+                result.append(sample[key])
+
+            eval_idx = 0
+            bn_adaption_executed = False
+            acc_within_tolerance = 0
+            pair_objective = None
+            in_cache = False
+            for evaluator in self._search._evaluators:
+                in_cache, value = evaluator.retrieve_from_cache(tuple(x[i]))
+                if not bn_adaption_executed and not in_cache:
+                    self._search._bn_adaptation.run(self._search._model)
+                    bn_adaption_executed = True
+                if evaluator.type_of_measurement == 'accuracy':
+                    if not in_cache:
+                        value, _, _ = evaluator.evaluate_model(self._search._model)
+                        evaluator.add_to_cache(tuple(x[i]), value)
+                    evaluators_arr[eval_idx].append(value * -1.0)
+                    upper_bound = 100
+                    lower_bound = self._search.search_params._ref_acc - self._search.acc_delta
+                    temp_acc = (value * -1.0) if value < 0 else value
+                    if temp_acc < upper_bound and temp_acc > lower_bound:
+                        acc_within_tolerance = temp_acc
+                else:
+                    if not in_cache:
+                        if evaluator.use_model_for_evaluation:
+                            value = evaluator.evaluate_model(self._search._model)
+                        else:
+                            value = evaluator.evaluate_with_elasticity_handler()
```

Comment:
Too branching logic. Most probably won't pass pylint because of this issue.

**Encapsulation of evaluators is broken.** 
evaluator has all information to evaluate `x` without external call `retrieve_from_cache` and checking for `in_cache` or checking `use_model_for_evaluation`.

It should be hidden inside the method of evaluator - `evaluate`.

If condition for `type_of_measurement` is not needed. The logic with list of evaluators should call API of Evaluator, instead of accessing its internals.
Keep in mind, that user may provide an external evaluator. Checking `type_of_measurement` won't make sense.

Reply:
I have refactored to reduce the exposure of internals of evaluators. Please take a look new approach. Thanks!